### PR TITLE
provisioner: add additional setup behavior

### PR DESCRIPTION
**Pull request by rkolchmeyer**
As part of setup, we need to download the build contexts, stop update
engine, and run docker-credential-gcr.

### CODE REVIEW

File: src/pkg/provisioner/provisioner.go [hunk 1]
```diff
@@ -17,24 +17,69 @@
...
 
+func setup(systemd *systemdClient) error {
+	log.Println("Setting up environment...")
+	if err := systemd.stop("update-engine.service"); err != nil {
+		return err
+	}
+	if err := unix.Mount("", "/root", "tmpfs", 0, ""); err != nil {
+		return fmt.Errorf("error mounting tmpfs at /root: %v", err)
+	}
+	cmd := exec.Command("docker-credential-gcr", "configure-docker")
```
Comment (arnav-kansal) [inline src/pkg/provisioner/provisioner.go]: should `docker-credential-gcr` also be part of the `Deps` structure?

File: src/pkg/provisioner/state.go [hunk 2]
```diff
@@ -52,7 +60,58 @@ func (s *state) write() error {
...
+		bucket, object := splitAddr[0], splitAddr[1]
+		tarPath := filepath.Join(s.dir, name+".tar")
+		if err := downloadGCSObject(ctx, deps.GCSClient, bucket, object, tarPath); err != nil {
+			return fmt.Errorf("error downloading %q to %q: %v", address, tarPath, err)
+		}
+		tarDir := filepath.Join(s.dir, name)
+		if err := os.Mkdir(tarDir, 0770); err != nil {
+			return err
+		}
+		args := []string{"xf", tarPath, "-C", tarDir}
```
Comment (arnav-kansal) [inline src/pkg/provisioner/state.go:100]: should some of these args come from `Deps.Tarcmd` somehow? or is the `Tarcmd` basically allowing supply of something like an alias to tar?

File: src/pkg/utils/utils.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"fmt"
+	"io"
+	"log"
+)
+
+// CheckClose closes an io.Closer and checks its error. Useful for checking the
+// errors on deferred Close() behaviors.
+func CheckClose(closer io.Closer, msg string, err *error) {
```
Comment (arnav-kansal) [inline src/pkg/utils/utils.go]: nit: maybe change `msg` to something like `errMsg`/ `additionalErrMsg`/ `errMsgOnClose` etc?

File: src/pkg/provisioner/provisioner.go (hunk 1 above)
Comment (rkolchmeyer) [reply src/pkg/provisioner/provisioner.go]: Done.

File: src/pkg/provisioner/state.go (hunk 2 above)
Comment (rkolchmeyer) [reply src/pkg/provisioner/state.go:100]: TarCmd is for providing a tar alias, yeah. It's helpful for testing, if you don't want to depend on the tar implementation in the environment. For now, I think the interface is cleaner without allowing for passing additional args, but we can change it later if it makes sense later.

File: src/pkg/utils/utils.go (hunk 3 above)
Comment (rkolchmeyer) [reply src/pkg/utils/utils.go]: Done.

File: src/pkg/provisioner/provisioner.go (hunk 1 above)
Comment (rkolchmeyer) [reply src/pkg/provisioner/provisioner.go]: Missed one spot in the rebase...this isn't done yet

Comment (rkolchmeyer) [reply src/pkg/provisioner/provisioner.go]: Now this is done.

**Review by arnav-kansal — APPROVED** [review]

**Pull request merged**